src/Microsoft.ML.Core/Data/Schema.cs

```diff
@@ -0,0 +1,344 @@
+    /// - Column name.
+    /// - Column type.
+    /// - Metadata. The metadata itself is a 'single-row dataset' (namely, an instance of <see cref="Metadata"/>), that contains its own schema
+    /// and values.
+    /// </summary>
+    public sealed class Schema : ISchema
+    {
+        private readonly Column[] _columns;
+        private readonly Dictionary<string, int> _nameMap;
+
+        /// <summary>
+        /// Number of columns in the schema.
+        /// </summary>
+        public int ColumnCount => _columns.Length;
+
+        /// <summary>
+        /// Get the column by name. Throws an exception if such column does not exist.
+        /// Note that if multiple columns exist with the same name, the one with the biggest index is returned.
+        /// The other columns are considered 'hidden', and only accessible by their index.
+        /// </summary>
+        public Column this[string name]
+        {
+            get
+            {
+                Contracts.CheckValue(name, nameof(name));
+                if (!_nameMap.TryGetValue(name, out int col))
+                    throw Contracts.ExceptParam(nameof(name), $"Column '{name}' not found");
+                return _columns[col];
+            }
+        }
+
+        /// <summary>
+        /// Get the column by index.
+        /// </summary>
+        public Column this[int col]
+        {
+            get
+            {
+                Contracts.CheckParam(0 <= col && col < _columns.Length, nameof(col));
+                return _columns[col];
+            }
+        }
+
+        /// <summary>
+        /// Retrieves an index of a column by name.
+        /// </summary>
+        /// <returns>True iff the column is present in the schema.</returns>
+        public bool TryGetColumnIndex(string name, out int col) => _nameMap.TryGetValue(name, out col);
+
+        /// <summary>
+        /// Get the column by name, or <c>null</c> if the column is not present.
+        /// </summary>
+        public Column GetColumnOrNull(string name)
+        {
+            Contracts.CheckNonEmpty(name, nameof(name));
+            if (_nameMap.TryGetValue(name, out int col))
+                return _columns[col];
+            return null;
+        }
+
+        /// <summary>
+        /// This class describes one column in the schema.
+        /// </summary>
+        public sealed class Column
+        {
+            /// <summary>
+            /// The name of the column.
+            /// </summary>
+            public string Name { get; }
+
+            /// <summary>
+            /// The type of the column.
+            /// </summary>
+            public ColumnType Type { get; }
+
+            /// <summary>
+            /// The metadata of the column.
+            /// </summary>
+            public Metadata Metadata { get; }
+
+            public Column(string name, ColumnType type, Metadata metadata)
```

Comment:
>metadata [](start = 65, length = 8)

any reason why you don't want to have default nullable here?

Reply:
I was thinking about it. For now, let's keep it as is and observe some more usage patterns.